### PR TITLE
refactor: replace next auth with lucia

### LINKED ISSUE #108: replace nextauth
I'm not happy with next auth. It does the session management, and magic link stuff well, but it's very restricting. I can't send a magic link from the server, and it doesn't work well when put on a different domain. I'm going to look in to alternatives. We really just need a library to handle the authentication and session management, and we can handle all of the UI, and necessary emails.

[Lucia](https://lucia-auth.com/tokens/start-here/introduction?) looks promising, but it's relatively new. @Pertempto what do you think?

### PR DESCRIPTION
**Pull request by arrocke**
## What has changed

* removed nextauth in favor of lucia
* added routes for login, logout, and getting the session
* added a view for login
* added emails for invite and login

~~I ran into one issue with lucia-auth. It doesn't support same site none cookies, which means that cookies won't work in vercel preview environments. `*.vercel.app` domains are considered different sites otherwise a cookie issued by one vercel app would be accessible in another. If we pay for vercel pro ($20/mo), then we have the option to pay an additional $100/mo to get our own domain in preview urls. We are at our limit of functions before we have to pay for vercel, so we need to make a decision for how we want to proceed. I'm ok forgoing preview URLs at the moment while we figure out a longer term hosting solution. We can have that conversation to [this issue](https://github.com/arrocke/gloss-translation/issues/112).~~

## Connected Issues

closes #108 

### Setup admin user (devyn61@ethereal.email)
1. run `nx prisma api migrate reset`
2. run `pg_restore -Fc --format=custom --dbname=<database> data/seed.dump`

### Login/Logout
1. Enter email on login view
4. Click link in sent email
5. Check that user is logged in
6. Click log out
7. Check that user is logged out

### Invite
1. Make sure you are logged in
2. Enter user info in the invite page
3. Click link in invite email
4. Check that new user is logged in.

## Post-Deployment

Will need to migrate existing users to new model

### CODE REVIEW

File: packages/api/shared/auth.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import { nextjs } from 'lucia-auth/middleware';
+import prisma from '@lucia-auth/adapter-prisma';
+import { client, ulid } from './db';
+import 'lucia-auth/polyfill/node';
+
+export const auth = lucia({
+  adapter: prisma(client),
+  env: process.env.NODE_ENV === 'production' ? 'PROD' : 'DEV',
+  middleware: nextjs(),
+  // TODO: extract to env vars
```
Comment (Pertempto) [inline packages/api/shared/auth.ts]: Does this need done in this PR?

File: packages/api/pages/api/auth/login.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        const token = randomBytes(12).toString('hex');
+        await auth.createKey(key.userId, {
+          type: 'single_use',
+          providerId: 'email-verification',
+          providerUserId: token,
+          password: null,
+          expiresIn: 60 * 60,
+        });
+
+        const url = new URL('http://localhost:4300/api/auth/login');
```
Comment (Pertempto) [inline packages/api/pages/api/auth/login.ts]: This won't work on production?

**Review by Pertempto — COMMENTED** [review]
I've started review on this. We're going on a camping trip this weekend, but hopefully I can continue on Monday.

**Comment by arrocke** [comment]
I addressed the environment variable issues, and found a way to get this working in preview URLs as well.

**Comment by Pertempto** [comment]
I logged in with the default account, then attempted to add an account. I got this error in the API log:
```
PrismaClientValidationError: 
Invalid `prisma.authKey.create()` invocation:
{
  data: {
    id: 'email-verification:133cad8ef033fd1801a9fcdd',
    user_id: undefined,
?   hashed_password?: String | null,
    primary_key: false,
    expires: 1687776739404,
+   auth_user: {
+     create?: AuthUserCreateWithoutAuth_keyInput | AuthUserUncheckedCreateWithoutAuth_keyInput,
+     connectOrCreate?: AuthUserCreateOrConnectWithoutAuth_keyInput,
+     connect?: AuthUserWhereUniqueInput
+   }
  }
}
Argument auth_user for data.auth_user is missing.
Note: Lines with + are required, lines with ? are optional.
    at ti.validate (webpack-internal:///(api)/./prisma/client/runtime/library.js:8634:37)     
    at Wr.createMessage (webpack-internal:///(api)/./prisma/client/runtime/library.js:10057:20)
    at eval (webpack-internal:///(api)/./prisma/client/runtime/library.js:11327:49)
    at runInChildSpan (webpack-internal:///(api)/./prisma/client/runtime/library.js:7117:59)  
    at t11._executeRequest (webpack-internal:///(api)/./prisma/client/runtime/library.js:11327:33) {
  clientVersion: '4.11.0'
}
```

The strange thing is, a new row was successfully added to the `User` table.

**Comment by Pertempto** [comment]
> I logged in with the default account, then attempted to add an account. I got this error in the API log:
>  ...

I added some `console.log` in `login.ts`

```
      if (key) {
        const token = randomBytes(12).toString('hex');
        console.log('Create key!');
        console.log(key);
        console.log(token);
        await auth.createKey(key.userId, {
          type: 'single_use',
          providerId: 'email-verification',
          providerUserId: token,
          password: null,
          expiresIn: 60 * 60,
        });
        console.log('success');
```

Here is the output
```
Create key!
{
  type: 'persistent',
  primary: true,
  providerId: 'username',
  providerUserId: 'devyn61@ethereal.email',
  userId: '0188d4d9-999d-26f9-a9df-e9df68c789b2',
  passwordDefined: false
}
0c979b5289806b17e607dc4b
success
PrismaClientValidationError: 
Invalid `prisma.authKey.create()` invocation:
{
  data: {
    id: 'email-verification:923d8b2e96b2e66e005ac06e',
    user_id: undefined,
?   hashed_password?: String | null,
    primary_key: false,
    expires: 1687819087652,
+   auth_user: {
+     create?: AuthUserCreateWithoutAuth_keyInput | AuthUserUncheckedCreateWithoutAuth_keyInput,
+     connectOrCreate?: AuthUserCreateOrConnectWithoutAuth_keyInput,
+     connect?: AuthUserWhereUniqueInput
+   }
  }
}
Argument auth_user for data.auth_user is missing.
Note: Lines with + are required, lines with ? are optional.
    at ti.validate (webpack-internal:///(api)/./prisma/client/runtime/library.js:8634:37)
    at Wr.createMessage (webpack-internal:///(api)/./prisma/client/runtime/library.js:10057:20)
    at eval (webpack-internal:///(api)/./prisma/client/runtime/library.js:11327:49)
    at runInChildSpan (webpack-internal:///(api)/./prisma/client/runtime/library.js:7117:59)
    at t11._executeRequest (webpack-internal:///(api)/./prisma/client/runtime/library.js:11327:33) {
  clientVersion: '4.11.0'
}
```

I don't understand the lucia key code very well, but it seems strange that the `providerUserId` is `devyn61@ethereal.email`. Should it be the email I am providing? In this case, `test@example.com`

**Comment by Pertempto** [comment]
Something is failing with the `auth.createKey` in `api/users/index.ts`, but I don't know enough about lucia to debug. The user is successfully created on line 51, but it fails on line 64.

**Review by Pertempto — CHANGES_REQUESTED** [review]
See my above comments about user invite bug.

**Comment by arrocke** [comment]
Should be good now

**Review by Pertempto — APPROVED** [review]
Works great!

**Pull request merged**